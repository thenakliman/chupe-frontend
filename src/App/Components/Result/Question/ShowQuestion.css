.show-question {
  position: absolute;
  margin-top: 4rem;
}