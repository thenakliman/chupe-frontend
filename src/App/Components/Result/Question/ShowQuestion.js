import React from 'react';
import propTypes from 'prop-types';
/* eslint-disable */
import {QuestionAnswerWrapper} from './QuestionAnswerWrapper';
import {QuestionDescription} from './QuestionDescription';
/* eslint-enable */
require('./ShowQuestion.css');

/** Question component */
export class ShowQuestion extends React.Component {
  /** Constructor defining the default values for the fields.
   * @param {Object} props, properties for the component.
   */
  constructor(props) {
    super(props);
    this.state = {
      question: '',
      description: '',
      owner: '',
      assignedTo: '',
      status: '',
      priority: '',
    };

    this.handleQuestionDescription = this
        .handleQuestionDescription.bind(this);

    this.handleAnswerDescription = this
        .handleAnswerDescription.bind(this);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.changeStatus = this.changeStatus.bind(this);
    this.changePriority = this.changePriority.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.getInitialState = this.getInitialState.bind(this);
    this.validateForm = this.validateForm.bind(this);
    this.addAnswer = this.addAnswer.bind(this);
  }

  /** Add answer
   * @param {string} answer, is the answer to the question.
   */
  addAnswer(answer) {
    const completeAnswer = {
      answer: answer,
      answeredBy: this.props.loggedInUser,
      questionId: this.props.match.params.id,
    };

    this.props.addAnswer(completeAnswer);
  }

  /** Provide the initial state to the component
   * @return {Object} initial state of the component
   */
  getInitialState() {
    const question = this.props.questions.find(
        (question) => question.id == this.props.match.params.id) || {};
    return {
      question: question.question,
      description: question.description,
      owner: question.owner,
      assignedTo: question.assignedTo,
      id: question.id,
      status: question.status,
      priority: question.priority,
    };
  }

  /** Set initial state on component mount */
  componentDidMount() {
    if (!this.props.questions.length) {
      this.props.getAllQuestions();
    }

    if (!this.props.users.length) {
      this.props.getAllUsers();
    }

    this.props.getAnswers(this.props.match.params.id);
    this.setState(this.getInitialState());
  }

  /** component did update
   *  @param {object} prevProps previous props
   */
  componentDidUpdate(prevProps) {
    /* todo: Error prone code, correct this code by explicitly checking
     * equality for each element of the array.
     */
    if (this.props.questions !== prevProps.questions) {
      this.setState(this.getInitialState());
    }
  }

  /** Handle changes on the question description fields
   * @param {object} description new modified value
   */
  handleQuestionDescription(description) {
    const completeQuestion = {
      id: this.state.id,
      question: this.state.question,
      description: description,
      assignedTo: this.state.assignedTo,
      owner: this.state.owner,
      status: this.state.status,
      priority: this.state.priority,
    };

    this.handleSubmit(completeQuestion);
  }

  /** Handle changes on the question description fields
   * @param {answerId} answerId to be modified
   * @param {object} description new modified value
   */
  handleAnswerDescription(answerId, description) {
    const completeAnswer = {
      id: answerId,
      answer: description,
      questionId: this.state.id,
      answeredBy: this.props.answers.find(
          (answer) => answer.id === answerId).answeredBy,
    };

    this.props.updateAnswer(answerId, completeAnswer);
  }

  /** Validates form data
   * @param {object} questionData has question data
   * @return {bool} return whether data is valid or not
   */
  validateForm(questionData) {
    return (
        !questionData.question ||
        !questionData.description ||
        !questionData.owner ||
        !questionData.assignedTo);
  }

  /** Handles submit of the form
   * @param {object} question containing new modified value
   */
  handleSubmit(question) {
    const hasError = this.validateForm(question);
    if (hasError) {
      return;
    }
    this.props.updateQuestion(this.props.questions, question);
  }

  /** Change status of the question
   * @param {string} status of the question
   */
  changeStatus(status) {
    this.handleSubmit(Object.assign({}, this.state, {status: status}));
  }

  /** Change priority of the question
   * @param {string} priority of the question
   */
  changePriority(priority) {
    this.handleSubmit(Object.assign({}, this.state, {priority: priority}));
  }

  /** Renders show question component
   @return {object} returns ShowQuestion component
   */
  render() {
    return (
        <div className='show-question'>
          <div className='question-description'>
            <QuestionDescription summary={this.state.question}
                                 owner={this.state.owner}
                                 priority={this.state.priority}
                                 status={this.state.status}
                                 id={this.state.id}
                                 changeStatus={this.changeStatus}
                                 changePriority={this.changePriority}
                                 assignedTo={this.state.assignedTo}/>
          </div>
          <div>
            <QuestionAnswerWrapper
                headerText={'Description Added by ' + this.state.owner}
                bodyText={this.state.description}
                id='question-answer-description-wrapper-id'
                saveHandler={this.handleQuestionDescription}
            />
          </div>
          {this.props.answers.map((answer) => (
              <div key={answer.id}>
                <QuestionAnswerWrapper
                    headerText={'Answer Added by ' + answer.answeredBy}
                    bodyText={answer.answer}
                    saveHandler={(description) =>
                        this.handleAnswerDescription(answer.id, description)}
                    id={`show-question--answer-${answer.id}-id`}
                />
              </div>
          ))}
          <div>
            <QuestionAnswerWrapper
                headerText={'Description Added by ' + this.props.loggedInUser}
                bodyText=''
                isEditing={true}
                id='show-question-answer--temporary-id'
                saveHandler={this.addAnswer}
            />
          </div>
        </div>
    );
  }
}

ShowQuestion.propTypes = {
  users: propTypes.arrayOf(propTypes.object).isRequired,
  questions: propTypes.arrayOf(propTypes.object).isRequired,
  answers: propTypes.arrayOf(propTypes.object).isRequired,
  loggedInUser: propTypes.string.isRequired,
  updateQuestion: propTypes.func.isRequired,
  getAnswers: propTypes.func.isRequired,
  addAnswer: propTypes.func.isRequired,
  getAllQuestions: propTypes.func.isRequired,
  getAllUsers: propTypes.func.isRequired,
  updateAnswer: propTypes.func.isRequired,
};
