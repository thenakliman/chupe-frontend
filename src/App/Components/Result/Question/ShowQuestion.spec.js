/* eslint-disable */
import React from 'react';
import {ShowQuestion} from './ShowQuestion';
/* eslint-enable */
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';

const questions = [{
    question: 'question1',
    id: 1,
    description: 'test-description1',
    owner: 'owner-1',
    assignedTo: 'assigned-1',
  },
  {
    question: 'question2',
    id: 2,
    description: 'test-description2',
    owner: 'owner-2',
    assignedTo: 'assigned-2',
  }];

const users = [{'userName': 'user1'}, {'userName': 'user2'}];

describe('Show Question component snapshot', () => {
    it('should match the snapshot', () => {
        const wrapper = shallow(
            <ShowQuestion
              users={users}
              isEditing={false}
              questions={questions}
              setEditingQuestion={()=>{}}
              updateQuestion={()=>{}}
              match={{params: {id: 2}}}
        />);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});


describe('Show question component', () => {
  it('should have question input field', () => {
    const wrapper = shallow(
        <ShowQuestion
          isEditing={false}
          users= {users}
          questions={questions}
          updateQuestion={()=>{}}
          setEditingQuestion={jest.fn()}
          match={{params: {id: 2}}}
        />
    );

    expect(wrapper.find(
        '#show-question-input-field-id').length).toEqual(1);
  });

  it('should have question description', () => {
    const wrapper = shallow(
        <ShowQuestion
          isEditing={false}
          users= {users}
          questions={questions}
          updateQuestion={()=>{}}
          match={{params: {id: 2}}}
          setEditingQuestion={jest.fn()} />
    );
    expect(wrapper.find(
        '#show-question-description-input-field-id').length).toEqual(1);
  });

  it('should have assigned to select field', () => {
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          isEditing={false}
          questions={questions}
          updateQuestion={()=>{}}
          match={{params: {id: 2}}}
          setEditingQuestion={jest.fn()} />
    );
    expect(wrapper.find(
        '#show-question-assigned-to-input-field-id').length).toEqual(1);
  });

  it('should have owner select field', () => {
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          isEditing={false}
          questions={questions}
          updateQuestion={()=>{}}
          setEditingQuestion={jest.fn()}
          match={{params: {id: 2}}}
        />
    );
    expect(wrapper.find(
        '#show-question-assigned-to-input-field-id').length).toEqual(1);
  });

  it('should have edit button', () => {
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          isEditing={false}
          questions={questions}
          updateQuestion={()=>{}}
          setEditingQuestion={jest.fn()}
          match={{params: {id: 2}}}
        />
    );
    expect(wrapper.find(
        '#show-question-edit-button-id').length).toEqual(1);
  });

  it('should dispatch action on click of edit button', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = mount(
        <ShowQuestion
          users= {users}
          isEditing={false}id
          questions={questions}
          updateQuestion={()=>{}}
          setEditingQuestion={mockSetEditingQuestion}
          match={{params: {id: 2}}}
        />
    );
    wrapper.find('#show-question-edit-button-id').simulate('submit');
    expect(wrapper.find('#show-question-edit-button-id').get(0).props.children)
      .toEqual('Edit');
    expect(mockSetEditingQuestion).toHaveBeenCalledWith(true);
  });

  it('should disable input fields on isEditing is false', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          isEditing={false}
          questions={questions}
          updateQuestion={()=>{}}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-input-field-id')
      .get(0).props.disabled).toEqual(true);

    expect(wrapper.find('#show-question-description-input-field-id')
      .get(0).props.disabled).toEqual(true);

    expect(wrapper.find('#show-question-owner-input-field-id')
      .get(0).props.disabled).toEqual(true);

    expect(wrapper.find('#show-question-assigned-to-input-field-id')
      .get(0).props.disabled).toEqual(true);
  });

  it('should enable input fields on isEditing is true', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          isEditing={true}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-input-field-id')
      .get(0).props.disabled).toEqual(false);

    expect(wrapper.find('#show-question-description-input-field-id')
      .get(0).props.disabled).toEqual(false);

    expect(wrapper.find('#show-question-owner-input-field-id')
      .get(0).props.disabled).toEqual(false);
    expect(wrapper.find('#show-question-assigned-to-input-field-id')
      .get(0).props.disabled).toEqual(false);
  });

  it('should have button name save if edit is disabled', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          questions={questions}
          updateQuestion={()=>{}}
          isEditing={true}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-edit-button-id')
      .get(0).props.children).toEqual('Save');
  });

  it('should have button name Edit if edit is enabled', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          users= {users}
          questions={questions}
          updateQuestion={()=>{}}
          isEditing={false}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-edit-button-id')
      .get(0).props.children).toEqual('Edit');
  });

  it('should show all users for assigned to if edit is enabled', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          isEditing={true}
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-assigned-to-input-field-id')
      .get(0).props.children.length).toEqual(2); questions;
  });

  it('should show all users for owner if edit is enabled', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          questions={questions}
          updateQuestion={()=>{}}
          isEditing={true}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion} />
    );
    expect(wrapper.find('#show-question-owner-input-field-id')
      .get(0).props.children.length).toEqual(2);
  });

  it('Component should have required questions properties', () => {
    const wrapper = shallow(
        <ShowQuestion
          isEditing={true}
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={()=>{}} />
    );
    expect(wrapper.state()).toEqual({
      id: 2,
      question: 'question2',
      description: 'test-description2',
      owner: 'owner-2',
      assignedTo: 'assigned-2',
    });
  });

  it('should not call submit if any field is empty', () => {
    const mockSetEditingQuestion = jest.fn();
    const wrapper = shallow(
        <ShowQuestion
          isEditing={true}
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={mockSetEditingQuestion}/>
    );
    wrapper.find('#show-question-edit-button-id').simulate('submit');
    expect(mockSetEditingQuestion).not.toHaveBeenCalled();
  });

  it('should set question text field', () => {
    const event = {target: {value: 'why oh why?'}};
    const container = mount(
        <ShowQuestion
          isEditing={true}
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={()=>{}}/>
    );
    container.find('#show-question-input-field-id')
      .simulate('change', event);

    container.find('#show-question-input-field-id').simulate('submit');
    expect(container.state()).toEqual({
       'assignedTo': 'assigned-2',
       'description': 'test-description2',
       'id': 2,
       'owner': 'owner-2',
       'question': event.target.value,
    });
  });

  it('should set question description text field', () => {
    const event = {target: {value: 'why oh why?'}};
    const container = mount(
        <ShowQuestion
          questions={questions}
          updateQuestion={()=>{}}
          isEditing={true}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={()=>{}}/>
    );
    container.find('#show-question-description-input-field-id')
      .simulate('change', event);

    container.find('#show-question-description-input-field-id')
      .simulate('submit');
    expect(container.state()).toEqual({
       'assignedTo': 'assigned-2',
       'description': event.target.value,
       'id': 2,
       'owner': 'owner-2',
       'question': 'question2',
    });
  });

  it('should set question assigned to select field', () => {
    const event = {target: {value: 'user1'}};
    const container = mount(
        <ShowQuestion
          isEditing={true}
          questions={questions}
          updateQuestion={()=>{}}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={()=>{}}/>
    );
    container.find('#show-question-assigned-to-input-field-id')
      .simulate('change', event);

    container.find('#show-question-assigned-to-input-field-id')
      .simulate('submit');
    expect(container.state()).toEqual({
       'assignedTo': event.target.value,
       'description': 'test-description2',
       'id': 2,
       'owner': 'owner-2',
       'question': 'question2',
    });
  });

  it('should set question owner select field', () => {
    const event = {target: {value: 'user1'}};
    const container = mount(
        <ShowQuestion
          questions={questions}
          updateQuestion={()=>{}}
          isEditing={true}
          users={users}
          match={{params: {id: 2}}}
          setEditingQuestion={()=>{}}/>
    );
    container.find('#show-question-owner-input-field-id')
      .simulate('change', event);

    container.find('#show-question-owner-input-field-id').simulate('submit');
    expect(container.state()).toEqual({
       'assignedTo': 'assigned-2',
       'description': 'test-description2',
       'id': 2,
       'owner': event.target.value,
       'question': 'question2',
    });
  });

    it('should dispatch add question on submit', () => {
      const mockUpdateQuestion = jest.fn();
      const container = mount(
          <ShowQuestion
            isEditing={true}
            users={users}
            match={{params: {id: 2}}}
            questions={questions}
            updateQuestion={mockUpdateQuestion}
            setEditingQuestion={()=>{}}
            />
      );

      container.find('#show-question-edit-button-id').simulate('submit');
      const expQuestion = {
          question: 'question2',
          description: 'test-description2',
          assignedTo: 'assigned-2',
          owner: 'owner-2',
          id: 2,
      };
      expect(mockUpdateQuestion).toHaveBeenCalledWith(questions, expQuestion);
    });

    it('should not dispatch add question on submit if editing is false', () => {
      const mockUpdateQuestion = jest.fn();
      const container = mount(
          <ShowQuestion
            isEditing={false}
            users={users}
            match={{params: {id: 2}}}
            questions={questions}
            setEditingQuestion={()=>{}}
            updateQuestion={mockUpdateQuestion}
            />
      );

      container.find('#show-question-edit-button-id').simulate('submit');

      expect(mockUpdateQuestion).not.toHaveBeenCalled();
    });

    it('should not dispatch add question if there are empty fields', () => {
      const mockUpdateQuestion = jest.fn();
      const questionWithEmptyFields = [{
          id: 2,
          description: '',
          question: '',
          assignedTo: '',
          owner: ''}];

      const container = mount(
          <ShowQuestion
            isEditing={false}
            users={users}
            match={{params: {id: 2}}}
            questions={questionWithEmptyFields}
            setEditingQuestion={()=>{}}
            updateQuestion={mockUpdateQuestion}
            />
      );

      container.find('#show-question-edit-button-id').simulate('submit');

      expect(mockUpdateQuestion).not.toHaveBeenCalled();
    });
});
