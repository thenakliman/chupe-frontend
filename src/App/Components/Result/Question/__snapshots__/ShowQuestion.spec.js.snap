// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Question component snapshot should match the snapshot 1`] = `
<div
  className="show-question"
>
  <p>
    <label
      htmlFor="show-question-input-field-id"
    >
      Question:
    </label>
    <input
      id="show-question-input-field-id"
      readOnly={true}
      type="text"
      value="q"
    />
  </p>
  <p>
    <label
      htmlFor="show-question-description-input-field-id"
    >
      Description:
    </label>
    <input
      id="show-question-description-input-field-id"
      readOnly={true}
      type="text"
      value="d"
    />
  </p>
  <p>
    <label>
      Assigned To:
    </label>
    <select
      id="show-question-assigned-to-input-field-id"
    >
      <option
        value=""
      >
        at
      </option>
    </select>
  </p>
  <p>
    <label>
      Owner:
    </label>
    <select
      id="show-question-owner-input-field-id"
    >
      <option
        value=""
      />
    </select>
  </p>
</div>
`;
